keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,409 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Many workloads benefit significantly from being aware of the topology of the node they are scheduled on.
+However, Kubernetes currently lacks a built-in mechanism to directly expose this node topology information to Pods.
+This KEP proposes a built-in admission plugin to copy standard node topology labels to Pods.
+Once copied, these labels can be consumed via the Downward API, just like any other Pod label.
+This approach eliminates the need for workarounds, such as custom init containers with elevated privileges,
+promoting a more secure and consistent solution.
+
+## Motivation
+
+Topology awareness is crucial for a growing number of applications and workloads on Kubernetes.
+Knowing the Pods's location within the cluster's topology allows for significant performance
+optimizations and improved resilience. End-user feedback has highlighted the following key use cases:
+* High-bandwidth AI/ML workloads, especially those employing distributed training or inference,
+  demonstrate a substantial need for topology awareness. The performance of these workloads is
+  highly sensitive to communication latency between GPUs. By preferentially scheduling Pods
+  requiring GPU-to-GPU communication within the same zone or rack, training and inference
+  times can be dramatically reduced.  Importantly, while Kubernetes can facilitate this topology-aware
+  placement, the actual orchestration of optimized GPU-to-GPU communication often relies on topology awareness
+  within the machine learning framework itself (e.g., Ray, PyTorch Distributed. etc).  Kubernetes provides the
+  necessary foundation for these frameworks to operate efficiently, but the frameworks are ultimately responsible
+  for leveraging that foundation.
+* CNI plugins can leverage topology information to establish more efficient network paths,
+  reducing latency and increasing throughput. For instance, a CNI could prioritize connections
+  within the same availability zone or rack.
+* Distributed stateful applications, such as sharded databases, can use topology information to improve fault tolerance.
+  By spreading replicas and traffic across different failure domains (e.g., zones, racks), these applications can achieve
+  higher availability and resilience to failures in any one topology.
+
+Today, applications obtain topology information by using an init container to query the Kubernetes API for the underlying node,
+extracting the necessary data from the Node resource. As more workloads and use cases benefit from topology awareness,
+we aim to simplify access to this information for Pods via the Downward API.
+
+### Goals
+
+* Values from Node labels `topology.k8s.io/zone`, `topology.k8s.io/region` and `kubernetes.io/hostname` are made
```

Comment:
Are you dropping the `*.topology.k8s.io` affordance from this KEP?

Reply:
On one hand, less scope is more gooder.

On the other hand, I thought it was expressly part of the original KEP that this be more extensible?  Yes, users can install a mutating admission - is that oo heavyweight for this?  How common will extensibility be?

I guess we can wait and see.

Reply:
Yes, I decided to only include the standard labels for now as it's safer and doing `*.topology.k8s.io/*` doesn't seem that much more extensible than what we doing now. Happy to revisit this after alpha once we get some user feedback.